### PR TITLE
feat(cli): Replace bundling with eszip in deno compile

### PR DESCRIPTION
**Pull request by williamtetlow**
https://github.com/denoland/deno/issues/13430

Replace bundling with `eszip` in `deno compile`.

Also adds support for dynamic imports in standalone executables https://github.com/denoland/deno/issues/8655

### CODE REVIEW

**Comment by bartlomieju** [comment]
@littledivy please take a look

**Comment by williamtetlow** [comment]
> @littledivy please take a look
@littledivy @bartlomieju this isn't ready for review.

I was making draft PR but it auto added you as reviewer.

I will tag you once it's ready for review

**Comment by williamtetlow** [comment]
@littledivy @bartlomieju this is ready for review

File: Cargo.lock [hunk 1]
```diff
@@ -1398,6 +1428,25 @@ dependencies = [
...
 
+[[package]]
+name = "eszip"
+version = "0.15.2"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "554c44bf4b0f383e601e884ffd3e48143a0523ff123f1fa7a0dba8d1041f469f"
+dependencies = [
+ "anyhow",
+ "base64 0.13.0",
+ "deno_ast 0.10.0",
```
Comment (bartlomieju) [inline Cargo.lock:1439]: Note to self: Need to update `eszip` before landing this PR

File: cli/main.rs [hunk 2]
```diff
@@ -1404,34 +1396,40 @@ pub fn main() {
...
-  let flags = match flags::flags_from_vec(args) {
-    Ok(flags) => flags,
-    Err(err @ clap::Error { .. })
-      if err.kind == clap::ErrorKind::DisplayHelp
-        || err.kind == clap::ErrorKind::DisplayVersion =>
-    {
-      err.print().unwrap();
-      std::process::exit(0);
+
+  let exit_code = async move {
```
Comment (bartlomieju) [inline cli/main.rs:1400]: This seems like unrelated change. I suggest to revert

File: cli/standalone.rs [hunk 3]
```diff
@@ -67,37 +64,51 @@ pub const MAGIC_TRAILER: &[u8; 8] = b"d3n0l4nd";
...
+  let (eszip, loader) = eszip::EszipV2::parse(bufreader)
+    .await
+    .context("Failed to parse eszip header")?;
+
+  let mut bufreader = loader.await.context("Failed to parse eszip archive")?;
+
+  bufreader.seek(SeekFrom::Start(metadata_pos)).await?;
+
+  let mut metadata = String::new();
+
```
Comment (bartlomieju) [inline cli/standalone.rs:101]: Very nice 👍

**Review by bartlomieju — COMMENTED** [review]
@williamtetlow looks good! Before landing we should ensure that stack traces behave correctly. Can you add an integration tests that compiles following script?
```
// a.ts
import "./b.ts";
```

```
// b.ts
console.log("hello");
throw new Error("boom!");
```

**Review by littledivy — CHANGES_REQUESTED** [review]
Unfortunately, JSON import assertions will no longer work.
```
~/g/deno (williamtetlow/replace-bundling-with-eszip-for-deno-compile)> target/release/deno compile -A ./a.ts
Check file:///Users/divy/gh/deno/a.ts
Compile file:///Users/divy/gh/deno/a.ts
Emit a
~/g/deno (williamtetlow/replace-bundling-with-eszip-for-deno-compile)> ./a
error: Expected a "JSON" module but loaded a "JavaScript" module.
```

File: cli/standalone.rs [hunk 4]
```diff
@@ -107,61 +118,44 @@ fn u64_from_bytes(arr: &[u8]) -> Result<u64, AnyError> {
...
       }
 
+      let code = module?.source().await;
+      let code = std::str::from_utf8(&code)
+        .map_err(|_| type_error("Module source is not utf-8"))?
+        .to_owned();
+
       Ok(deno_core::ModuleSource {
         code,
         module_type: deno_core::ModuleType::JavaScript,
```
Comment (littledivy) [inline cli/standalone.rs:161]:
```suggestion
        module_type: match module.kind {
          eszip::ModuleKind::JavaScript => deno_core::ModuleType::JavaScript,
          eszip::ModuleKind::Json => deno_core::ModuleType::Json,
        },
```

**Comment by williamtetlow** [comment]
I think this change will allow us to support dynamic imports in standalone executables too https://github.com/denoland/deno/pull/13563/commits/f2b55e86f4dc23fbfe9264d77843c197c5c40a8a

File: cli/main.rs (hunk 2 above)
Comment (williamtetlow) [reply cli/main.rs:1400]: @bartlomieju this was done as both `extract_standalone` and `standalone::run` have been converted to async functions.

I wrapped the calls to these and also the call to the async fn `get_subcommand`  in an async block so we call `run_basic` only once.

File: cli/tests/integration/compile_tests.rs [hunk 5]
```diff
@@ -202,6 +208,38 @@ fn standalone_load_datauri() {
   assert_eq!(output.stdout, b"Hello Deno!\n");
 }
 
+#[test]
+fn standalone_load_dynamic_import() {
+  let dir = TempDir::new().expect("tempdir fail");
+  let exe = if cfg!(windows) {
```
Comment (littledivy) [inline cli/tests/integration/compile_tests.rs:214]: @williamtetlow Can we revert this change? eszip doesn't include dynamic imports in the graph as of 0.16.0 The loader should handle this instead and I'd prefer to do this in a follow up PR instead.

File: Cargo.lock [hunk 6]
```diff
@@ -488,6 +488,16 @@ dependencies = [
  "unicode-width",
 ]
 
+[[package]]
+name = "console_error_panic_hook"
+version = "0.1.7"
```
Comment (littledivy) [inline Cargo.lock:493]: Waiting on https://github.com/denoland/deno_doc/pull/245

**Review by littledivy — APPROVED** [review]
LGTM! Thanks @williamtetlow 

@bartlomieju PTAL

**Review by bartlomieju — APPROVED** [review]
I'm still not happy about changes to `cli/main.rs` but I can live with them and refactor them later. 

LGTM, nice work @williamtetlow

**Comment by vectronic** [comment]
@littledivy @williamtetlow 

I'm curious as to what changed in this PR that meant you had to remove this test:

https://github.com/denoland/deno/pull/13563/commits/0ff2ce00462416b30a04c66399c32597eab18d36

I assume you originally had this working due to this comment:

https://github.com/denoland/deno/pull/13563#issuecomment-1030240156
